### PR TITLE
Add vertex_state operation tests for all GPUVertexState params

### PR DESCRIPTION
**Pull request by Kangz**
This doesn't implement the entirety of the test plan yet. The fixture
is big enough to be in its own commit along with some tests. The rest of
the test plan will be implemented in a follow-up (with potential fixture
changes if needed).

<hr>

**Author checklist for test code/plans:**

- [X] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [X] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [x] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [x] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [x] Existing (or new) test helpers are used where they would reduce complexity.
- [x] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 1]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+g.test('setVertexBufferOffset_and_attributeOffset')
+  .desc(
+    `Test that the vertex buffer offset and attribute offset in the vertex state are applied correctly. Test for:
+  - all formats
+  - various setVertexBuffer offsets
+  - various attribute offsets in a fixed arrayStride`
+  )
+  .params(u =>
+    u //
+      .combine('format', kVertexFormats)
```
Comment (kainino0x) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:659]: optional: perhaps a .beginSubcases() here or somewhere around here?

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 2]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+    `Test that the raw data passed in vertex buffers is correctly converted to the input type in the shader. Test for:
+  - all formats
+  - 1 to 4 components in the shader's input type (unused components are filled with 0 and except the 4th with 1)
+  - various locations
+  - various slots`
+  )
+  .params(u =>
+    u //
+      .combine('format', kVertexFormats)
+      .combine('shaderComponentCount', [1, 2, 3, 4])
```
Comment (kainino0x) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:627]: nit: perhaps a .beginSubcases() here?

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 3]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+        ['instance'],
+        ['vertex', 'vertex', 'instance'],
+        ['instance', 'vertex', 'instance'],
+        ['vertex', 'instance', 'vertex', 'vertex'],
+      ] as const)
+  )
+  .fn(t => {
+    const { stepModes } = t.params;
+    const state = stepModes.map((stepMode, i) => {
+      return {
```
Comment (kainino0x) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:761]: You can avoid the return by doing `(stepMode, i) => ({ slot: i, etc })`

**Review by kainino0x — APPROVED** [review]
LGTM, very nice tests! Took a little bit to work through the code, but only because I couldn't focus. It was actually relatively easy to have a comprehensive understanding of the code.

Also I like that I can review an entire test by just looking at the params. :P

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 4]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+  expandTestData(data: TestData, maxIndex: number, componentCount: number): TestData {
+    const vertexComponentSize = data.vertexData.byteLength / data.testComponentCount;
+    const expectedComponentSize = data.expectedData.byteLength / data.testComponentCount;
+
+    const expandedVertexData = new Uint8Array(maxIndex * componentCount * vertexComponentSize);
+    const expandedExpectedData = new Uint8Array(4 * maxIndex * expectedComponentSize);
+
+    for (let index = 0; index < maxIndex; index++) {
+      for (let component = 0; component < componentCount; component++) {
+        // If only we had some builtin JS memcpy function between ArrayBuffers...
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:471]: I might actually spend the time to make an ECMAScript proposal ^^ CC @beaufortfrancois

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 5]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+      }
+    >,
+    maxVertexIndex: number,
+    maxInstanceIndex: number
+  ): string {
+    let vsInputs = '';
+    let vsChecks = '';
+    let vsBindings = '';
+
+    buffers.forEach(b => {
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:131]: TIL

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 6]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+        assert(formatInfo.bytesPerComponent <= 16);
+        let vertexData: ArrayBuffer | undefined = undefined;
+        if (formatInfo.bytesPerComponent === 2) {
+          vertexData = new Int16Array(data).buffer;
+        } else if (formatInfo.bytesPerComponent === 1) {
+          vertexData = new Int8Array(data).buffer;
+        }
+
+        const divider = Math.pow(2, formatInfo.bytesPerComponent * 8 - 1) - 1;
+        const expectedData = new Float32Array(data.map(v => Math.max(-1, v / divider))).buffer;
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:417]: Done also introduced a bitSize early in the function and replaced all formatInfo.bytesPerComponent * 8 with it.

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 7]
```diff
@@ -16,35 +16,760 @@ TODO: check overlap with api,operation,rendering,draw:vertex_attributes,basic be
...
+          testComponentCount: data.length,
+          expectedData,
+          vertexData: vertexData!,
+          floatTolerance: 0.1 / divider,
+        };
+      }
+    }
+  }
+
+  // The TestData generate for a format might not contain enough data for all the vertices we are
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:458]: I decided to skip this nit to not have to reflow all the comments D:

**Comment by Kangz** [comment]
Addressed all comments, PTAL again.

**Review by kainino0x — APPROVED** [review]

**Pull request merged**